<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#STARTUP-FUNDING-ANALYSIS:-Exploring-Venture-Capital-Investments-in-Various-Industries" data-toc-modified-id="STARTUP-FUNDING-ANALYSIS:-Exploring-Venture-Capital-Investments-in-Various-Industries-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><span style="undefined:undefined">STARTUP FUNDING ANALYSIS:</span> <strong><span style="undefined:undefined">Exploring Venture Capital Investments in Various Industries</span></strong></a></span></li><li><span><a href="#BUSINESS-UNDERSTANDING" data-toc-modified-id="BUSINESS-UNDERSTANDING-2"><span class="toc-item-num">2&nbsp;&nbsp;</span><span style="undefined:undefined">BUSINESS UNDERSTANDING</span></a></span><ul class="toc-item"><li><span><a href="#Background-information" data-toc-modified-id="Background-information-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Background information</a></span></li><li><span><a href="#Problem-Statement" data-toc-modified-id="Problem-Statement-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Problem Statement</a></span></li><li><span><a href="#Objectives" data-toc-modified-id="Objectives-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Objectives</a></span></li><li><span><a href="#Dataset-description." data-toc-modified-id="Dataset-description.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Dataset description.</a></span><ul class="toc-item"><li><span><a href="#Key-points:" data-toc-modified-id="Key-points:-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span><code>Key points:</code></a></span></li></ul></li></ul></li><li><span><a href="#DATA-UNDERSTANDING" data-toc-modified-id="DATA-UNDERSTANDING-3"><span class="toc-item-num">3&nbsp;&nbsp;</span><span style="undefined:undefined">DATA UNDERSTANDING</span></a></span><ul class="toc-item"><li><span><a href="#Import-libraries-and-modules" data-toc-modified-id="Import-libraries-and-modules-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Import libraries and modules</a></span></li><li><span><a href="#Create-a-dataframe-by-reading-data-from-a-csv-file.-Examining-the-first-five-rows" data-toc-modified-id="Create-a-dataframe-by-reading-data-from-a-csv-file.-Examining-the-first-five-rows-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Create a dataframe by reading data from a csv file. Examining the first five rows</a></span></li><li><span><a href="#Summary-of-Features-in-the-Dataset" data-toc-modified-id="Summary-of-Features-in-the-Dataset-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Summary of Features in the Dataset</a></span></li><li><span><a href="#class-to-describe-dataset" data-toc-modified-id="class-to-describe-dataset-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>class to describe dataset</a></span></li><li><span><a href="#Information-about-our-dataset" data-toc-modified-id="Information-about-our-dataset-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Information about our dataset</a></span></li><li><span><a href="#Check-shape-of-dataframe" data-toc-modified-id="Check-shape-of-dataframe-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Check shape of dataframe</a></span></li><li><span><a href="#checking-for-the-categorical-columns" data-toc-modified-id="checking-for-the-categorical-columns-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>checking for the categorical columns</a></span></li><li><span><a href="#checking-for-the-ccontinuous--columns" data-toc-modified-id="checking-for-the-ccontinuous--columns-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>checking for the ccontinuous  columns</a></span></li><li><span><a href="#Concise-statistical-description-of-numeric-features" data-toc-modified-id="Concise-statistical-description-of-numeric-features-3.9"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>Concise statistical description of numeric features</a></span></li><li><span><a href="#Checking-for-duplicated-rows" data-toc-modified-id="Checking-for-duplicated-rows-3.10"><span class="toc-item-num">3.10&nbsp;&nbsp;</span>Checking for duplicated rows</a></span></li><li><span><a href="#Checking-for-missing-values" data-toc-modified-id="Checking-for-missing-values-3.11"><span class="toc-item-num">3.11&nbsp;&nbsp;</span>Checking for missing values</a></span></li><li><span><a href="#Checking-and-Visualizing-outliers" data-toc-modified-id="Checking-and-Visualizing-outliers-3.12"><span class="toc-item-num">3.12&nbsp;&nbsp;</span>Checking and Visualizing outliers</a></span></li></ul></li><li><span><a href="#DATA-CLEANING" data-toc-modified-id="DATA-CLEANING-4"><span class="toc-item-num">4&nbsp;&nbsp;</span><span style="undefined:undefined">DATA CLEANING</span></a></span><ul class="toc-item"><li><span><a href="#Renamimg-column(s)-to-remove-spaces." data-toc-modified-id="Renamimg-column(s)-to-remove-spaces.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Renamimg column(s) to remove spaces.</a></span></li><li><span><a href="#dropping-duplicated-rows." data-toc-modified-id="dropping-duplicated-rows.-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>dropping duplicated rows.</a></span></li><li><span><a href="#Type-Casting:-Converting-Funding-Total-to-Float" data-toc-modified-id="Type-Casting:-Converting-Funding-Total-to-Float-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Type Casting: Converting Funding Total to Float</a></span></li><li><span><a href="#dropping-missing-values." data-toc-modified-id="dropping-missing-values.-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>dropping missing values.</a></span></li><li><span><a href="#Addressing-Missing-Values-for-Specific-Columns" data-toc-modified-id="Addressing-Missing-Values-for-Specific-Columns-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Addressing Missing Values for Specific Columns</a></span><ul class="toc-item"><li><span><a href="#Confirming-there-are-no-missing-values" data-toc-modified-id="Confirming-there-are-no-missing-values-4.5.1"><span class="toc-item-num">4.5.1&nbsp;&nbsp;</span>Confirming there are no missing values</a></span></li></ul></li><li><span><a href="#Dealing-with-outliers" data-toc-modified-id="Dealing-with-outliers-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Dealing with outliers</a></span></li></ul></li><li><span><a href="#EXPLORATORY-DATA-ANALYSIS(EDA)" data-toc-modified-id="EXPLORATORY-DATA-ANALYSIS(EDA)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span><span style="undefined:undefined">EXPLORATORY DATA ANALYSIS(EDA)</span></a></span><ul class="toc-item"><li><span><a href="#-Univariate-Analysis.--" data-toc-modified-id="-Univariate-Analysis.---5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span><span style="undefined:undefined"> Univariate Analysis.  </span></a></span><ul class="toc-item"><li><span><a href="#Analysis-on-status-Feature" data-toc-modified-id="Analysis-on-status-Feature-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Analysis on <code>status</code> Feature</a></span></li><li><span><a href="#top-10%-total_funding-startups.¶" data-toc-modified-id="top-10%-total_funding-startups.¶-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>top 10% <code>total_funding</code> startups.¶</a></span></li><li><span><a href="#What-is-the-distribution-of-founded-years?" data-toc-modified-id="What-is-the-distribution-of-founded-years?-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>What is the distribution of founded years?</a></span></li><li><span><a href="#What-are-the-Top-10-Country-Code-Counts?" data-toc-modified-id="What-are-the-Top-10-Country-Code-Counts?-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>What are the Top 10 Country Code Counts?</a></span></li><li><span><a href="#What-are-the-Top-10-City-Code-Counts?" data-toc-modified-id="What-are-the-Top-10-City-Code-Counts?-5.1.5"><span class="toc-item-num">5.1.5&nbsp;&nbsp;</span>What are the Top 10 City Code Counts?</a></span></li><li><span><a href="#What-is-the-distribution-of-funding-rounds?" data-toc-modified-id="What-is-the-distribution-of-funding-rounds?-5.1.6"><span class="toc-item-num">5.1.6&nbsp;&nbsp;</span>What is the distribution of funding rounds?</a></span></li><li><span><a href="#What-are-the-top-10-markets-based-on-count?" data-toc-modified-id="What-are-the-top-10-markets-based-on-count?-5.1.7"><span class="toc-item-num">5.1.7&nbsp;&nbsp;</span>What are the top 10 markets based on count?</a></span></li><li><span><a href="#'Markets-that-most-start-ups-closed" data-toc-modified-id="'Markets-that-most-start-ups-closed-5.1.8"><span class="toc-item-num">5.1.8&nbsp;&nbsp;</span>'Markets that most start-ups closed</a></span></li><li><span><a href="#Markets-that-least-start-ups-closed" data-toc-modified-id="Markets-that-least-start-ups-closed-5.1.9"><span class="toc-item-num">5.1.9&nbsp;&nbsp;</span>Markets that least start-ups closed</a></span></li></ul></li><li><span><a href="#-Multivariate-Analysis.--" data-toc-modified-id="-Multivariate-Analysis.---5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span><span style="undefined:undefined"> Multivariate Analysis.  </span></a></span><ul class="toc-item"><li><span><a href="#What-are-the-funding-totals-by-country?" data-toc-modified-id="What-are-the-funding-totals-by-country?-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>What are the funding totals by country?</a></span></li><li><span><a href="#What-are-the-top-10-funding-totals-by-founded-year?" data-toc-modified-id="What-are-the-top-10-funding-totals-by-founded-year?-5.2.2"><span class="toc-item-num">5.2.2&nbsp;&nbsp;</span>What are the top 10 funding totals by founded year?</a></span></li><li><span><a href="#What-is-the-average-funding-rounds-by-status?" data-toc-modified-id="What-is-the-average-funding-rounds-by-status?-5.2.3"><span class="toc-item-num">5.2.3&nbsp;&nbsp;</span>What is the average funding rounds by status?</a></span></li><li><span><a href="#What-is-the-relationship-between-funding-total-and-funding-rounds-by-market?" data-toc-modified-id="What-is-the-relationship-between-funding-total-and-funding-rounds-by-market?-5.2.4"><span class="toc-item-num">5.2.4&nbsp;&nbsp;</span>What is the relationship between funding total and funding rounds by market?</a></span></li><li><span><a href="#What-is-the-total-funding-received-by-companies?" data-toc-modified-id="What-is-the-total-funding-received-by-companies?-5.2.5"><span class="toc-item-num">5.2.5&nbsp;&nbsp;</span>What is the total funding received by companies?</a></span></li></ul></li></ul></li><li><span><a href="#FEATURE-ENGINEERING/SELECTION" data-toc-modified-id="FEATURE-ENGINEERING/SELECTION-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><span style="undefined:undefined">FEATURE ENGINEERING/SELECTION</span></a></span></li><li><span><a href="#MODELLING" data-toc-modified-id="MODELLING-7"><span class="toc-item-num">7&nbsp;&nbsp;</span><span style="undefined:undefined">MODELLING</span></a></span></li><li><span><a href="#EVALUATION" data-toc-modified-id="EVALUATION-8"><span class="toc-item-num">8&nbsp;&nbsp;</span><span style="undefined:undefined">EVALUATION</span></a></span></li><li><span><a href="#RECOMMENDATIONS-" data-toc-modified-id="RECOMMENDATIONS--9"><span class="toc-item-num">9&nbsp;&nbsp;</span><span style="undefined:undefined">RECOMMENDATIONS </span></a></span></li><li><span><a href="#CONCLUSION" data-toc-modified-id="CONCLUSION-10"><span class="toc-item-num">10&nbsp;&nbsp;</span><span style="undefined:undefined">CONCLUSION</span></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#The-Top-Countries-in-Terms-of-Startup-Number" data-toc-modified-id="The-Top-Countries-in-Terms-of-Startup-Number-10.0.1"><span class="toc-item-num">10.0.1&nbsp;&nbsp;</span>The Top Countries in Terms of Startup Number</a></span></li><li><span><a href="#most-funded-categoriess" data-toc-modified-id="most-funded-categoriess-10.0.2"><span class="toc-item-num">10.0.2&nbsp;&nbsp;</span>most funded categoriess</a></span></li><li><span><a href="#Top-10-Market-of-Start-ups" data-toc-modified-id="Top-10-Market-of-Start-ups-10.0.3"><span class="toc-item-num">10.0.3&nbsp;&nbsp;</span>Top 10 Market of Start-ups</a></span></li><li><span><a href="#What-is-the-relationship-between-funding-and-founded-year?-(With-points-colored-by-status)" data-toc-modified-id="What-is-the-relationship-between-funding-and-founded-year?-(With-points-colored-by-status)-10.0.4"><span class="toc-item-num">10.0.4&nbsp;&nbsp;</span>What is the relationship between funding and founded year? (With points colored by status)</a></span></li><li><span><a href="#The-Companies-get-more-than-10-Million-Dollars-by-angel-investors" data-toc-modified-id="The-Companies-get-more-than-10-Million-Dollars-by-angel-investors-10.0.5"><span class="toc-item-num">10.0.5&nbsp;&nbsp;</span>The Companies get more than 10 Million Dollars by angel investors</a></span></li><li><span><a href="#Number-of-Startups-Country-wise-on-Map" data-toc-modified-id="Number-of-Startups-Country-wise-on-Map-10.0.6"><span class="toc-item-num">10.0.6&nbsp;&nbsp;</span>Number of Startups Country wise on Map</a></span></li></ul></li></ul></li></ul></div>

 <h1 style="text-align: center;"><span style="color: #FF0000;">STARTUP FUNDING ANALYSIS:</span> <strong><span style="color: #0000FF;">Exploring Venture Capital Investments in Various Industries</span></strong></h1>


![Startup Image](startupimage2.jpeg)

**We would like to acknowledge the contributions of the team members:**

- Florence Nguuni
- Brian Chabari
- Diana Mwaura
- Isaac Muturi
- Kennedy Juma
- Joel Omondi


*******

# <span style="color: #008000;">BUSINESS UNDERSTANDING</span>



## Background information 

In today's rapidly evolving business landscape, the allure of founding a startup with a groundbreaking idea has captivated the imagination of many aspiring entrepreneurs. However, without a firm grasp of the past and insights from previous startup experiences, venturing into this exhilarating journey can be perilous. The dataset used in this kernel provides a unique opportunity to delve into the world of startups and gain a comprehensive understanding of their dynamics, challenges, and triumphs.

By analyzing various aspects such as founding year, funding received, and current operating status, we can uncover patterns and trends that contribute to startup success or failure. This knowledge can guide aspiring entrepreneurs and investors in making informed decisions, formulating robust strategies, and identifying untapped opportunities for growth. By studying the past of the startup culture, we can shape the future of startups, positioning them for increased performance and success in today's markets.

Investors face the challenge of identifying startups with high potential for success, often referred to as "unicorns," among a large pool of potential investments.
Various attempts have been made to predict the success of startups, including perspectives shared by investors and insights derived from machine learning and big data analysis.

However, these methods have limitations. Investor perspectives are valuable but difficult to replicate, and their metrics for evaluating founders' potential may not be universally applicable.
Data-driven approaches utilizing machine learning and artificial intelligence hold promise for predicting startup success. However, existing methods typically rely on publicly available data, which introduces its own set of assumptions and uncertainties.
The focus of the post will be on the data-driven approach, providing an overview of the dataset, standard AI practices, and a solution specifically tailored for predicting startup success.

## Problem Statement

Despite the rapid growth of the startup scene and the increasing investments from venture capital firms, understanding the dynamics and characteristics of these newly established firms remains a challenge. The chosen dataset provides a valuable glimpse into this captivating realm, presenting an opportunity to delve deeper and unravel the underlying trends and patterns within the startup ecosystem.

The problem at hand is to analyze the dataset and extract meaningful insights to gain a comprehensive understanding of the startup landscape, including factors such as funding trends, industry preferences, geographical distribution, and success metrics. By addressing this problem, we aim to unlock valuable knowledge that can guide aspiring entrepreneurs, investors, and policy-makers in making informed decisions and fostering the growth of the startup ecosystem.

## Objectives

1. **Identify Key Success Factors**: Determine the primary factors contributing to startup success by analyzing prevalent characteristics, strategies, and patterns among successful ventures.

2. **Optimize Funding Strategies**: Explore different funding patterns and sources to maximize funding success rate and amount received by startups.

3. **Mitigate Risks and Challenges**: Develop strategies to mitigate common pitfalls and challenges faced by startups, enhancing resilience and increasing survival and growth rates.

4. **Identify Market Opportunities**: Analyze industry sectors, market trends, and emerging niches to target untapped opportunities for market penetration and growth.

5. **Enhance Decision-Making**: Provide data-driven insights to support informed decision-making, optimizing choices related to market entry, resource allocation, and strategic partnerships.

## Dataset description.

* The dataset used for training the machine learning model contains investment data about startup companies, sourced from Crunchbase, a prominent website for company insights ranging from early-stage startups to Fortune 1000 companies.
* The dataset comprises approximately 54,000 rows and 39 columns, providing information about the companies, including their names, URLs, markets, countries, states, regions, cities, founded dates, and funding details.
* The funding details cover various types such as seed funding, venture equity crowdfunding, convertible notes, debt financing, angel investments, grants, private equity, post IPO equity, post IPO debt, secondary market, product crowdfunding, and round A-H series funding.
* Additionally, the dataset includes the status of the companies, categorized as acquired, operating, or closed.

### `Key points:`

* Dataset: Investment data comprising details of startup companies.

* Funding Rounds: Different funding rounds and types, including seed funding, venture equity crowdfunding, and convertible notes.

* Company Status: Classification of companies as acquired, operating, or closed.

* Challenge: Predicting startup success is difficult but valuable for investors.

* Limitations: Investor perspectives are subjective, and publicly available data has its own assumptions.

* Data-Driven Approach: Utilizing machine learning and AI to predict startup success.
* Solution Focus: Providing a solution developed specifically for this purpose.

# <span style="color: #008000;">DATA UNDERSTANDING</span>

## Import libraries and modules

In [1]:
# Import modules & packages

# Data manipulation
import pandas as pd
import numpy as np

# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Modeling
from sklearn.model_selection import train_test_split,cross_val_score,GridSearchCV #splitting the dataset into test-train
from imblearn.over_sampling import SMOTE
from sklearn.metrics import accuracy_score,f1_score,recall_score,precision_score,confusion_matrix,roc_curve,roc_auc_score,classification_report # performance metrics
from sklearn.preprocessing import MinMaxScaler, LabelEncoder, StandardScaler
from scipy import stats

# Feature Selection, Feature Importance
from sklearn.inspection import permutation_importance


# Algorithms for supervised learning methods
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

# Filtering future warnings
import warnings
warnings.filterwarnings('ignore')

## Create a dataframe by reading data from a csv file. Examining the first five rows

In [2]:
# Loading and previewing data
data = pd.read_csv('investments_VC.csv',encoding = 'latin')
data.head()

permalink                name  \
0            /organization/waywire            #waywire   
1  /organization/tv-communications  &TV Communications   
2    /organization/rock-your-paper   'Rock' Your Paper   
3   /organization/in-touch-network   (In)Touch Network   
4   /organization/r-ranch-and-mine  -R- Ranch and Mine   

                    homepage_url  \
0         http://www.waywire.com   
1          http://enjoyandtv.com   
2   http://www.rockyourpaper.org   
3  http://www.InTouchNetwork.com   
4                            NaN   

                                       category_list        market   \
0         |Entertainment|Politics|Social Media|News|          News    
1                                            |Games|         Games    
2                             |Publishing|Education|    Publishing    
3  |Electronics|Guides|Coffee|Restaurants|Music|i...   Electronics    
4                      |Tourism|Entertainment|Games|       Tourism    

   funding_total_usd      status country_code state_code         region  ...  \
0          17,50,000    acquired          USA         NY  New York City  ...   
1          40,00,000   operating          USA         CA    Los Angeles  ...   
2             40,000   operating          EST        NaN        Tallinn  ...   
3          15,00,000   operating          GBR        NaN         London  ...   
4             60,000   operating          USA         TX         Dallas  ...   

  secondary_market  product_crowdfunding round_A round_B round_C  round_D  \
0              0.0                   0.0     0.0     0.0     0.0      0.0   
1              0.0                   0.0     0.0     0.0     0.0      0.0   
2              0.0                   0.0     0.0     0.0     0.0      0.0   
3              0.0                   0.0     0.0     0.0     0.0      0.0   
4              0.0                   0.0     0.0     0.0     0.0      0.0   

  round_E round_F  round_G  round_H  
0     0.0     0.0      0.0      0.0  
1     0.0     0.0      0.0      0.0  
2     0.0     0.0      0.0      0.0  
3     0.0     0.0      0.0      0.0  
4     0.0     0.0      0.0      0.0  

[5 rows x 39 columns]

## Summary of Features in the Dataset

1. **permalink**: The unique identifier or URL of the company.
2. **name**: The name of the startup company.
3. **homepage_url**: The URL of the startup's website.
4. **category_list**: The categories or industries associated with the startup.
5. **market**: The market or industry segment in which the startup operates.
6. **funding_total_usd**: The total amount of funding raised by the startup in US dollars.
7. **status**: The status of the startup (e.g., active, acquired, closed).
8. **country_code**: The country code of the startup's headquarters or primary operations.
9. **state_code**: The state code of the startup's headquarters or primary operations.
10. **region**: The region where the startup is located.
11. **city**: The city where the startup is located.
12. **funding_rounds**: The number of funding rounds the startup has gone through.
13. **founded_at**: The specific date on which the startup was founded.
14. **founded_month**: The month in which the startup was founded.
15. **founded_quarter**: The quarter in which the startup was founded.
16. **founded_year**: The year in which the startup was founded.
17. **first_funding_at**: The date of the first funding round.
18. **last_funding_at**: The date of the most recent funding round.
19. **seed**: The amount of funding raised in the seed round.
20. **venture**: The amount of funding raised in the venture round.
21. **equity_crowdfunding**: The amount of funding raised through equity crowdfunding.
22. **undisclosed**: The amount of undisclosed funding raised.
23. **convertible_note**: The amount of funding raised through convertible notes.
24. **debt_financing**: The amount of funding raised through debt financing.
25. **angel**: The amount of funding raised from angel investors.
26. **grant**: The amount of grant funding received.
27. **private_equity**: The amount of funding raised from private equity investors.
28. **post_ipo_equity**: The amount of funding raised post-IPO through equity.
29. **post_ipo_debt**: The amount of funding raised post-IPO through debt.
30. **secondary_market**: The amount of funding raised through secondary market transactions.
31. **product_crowdfunding**: The amount of funding raised through product crowdfunding.
32. **round_A**: The amount of funding raised in round A.
33. **round_B**: The amount of funding raised in round B.
34. **round_C**: The amount of funding raised in round C.
35. **round_D**: The amount of funding raised in round D.
36. **round_E**: The amount of funding raised in round E.
37. **round_F**: The amount of funding raised in round F.
38. **round_G**: The amount of funding raised in round G.
39. **round_H**: The amount of funding raised in round H.

## class to describe dataset

In [3]:
# class to describe dataset

class Description:
    
    # initialize object
    def __init__(self, data):
        self.data = data
        
    # method to check shape of data
    def data_shape(self):
        out = print(f"The DataFrame has:\n\t* {self.data.shape[0]} rows\n\t* {self.data.shape[1]} columns", '\n')
        return out
    
    # method to check info on dataset
    def data_info(self):
        out = print(self.data.info(), '\n')
        return out
    
    # method to describe numerical columns
    def data_describe(self):
        out = self.data.describe()
        return out
    
    def run(self):
        self.data_shape()
        self.data_info()
        self.data_describe()

## Information about our dataset

In [4]:
# Instantiating the class
desc = Description(data)

# Checking the data types of the columns
desc.data_info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54294 entries, 0 to 54293
Data columns (total 39 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   permalink             49438 non-null  object 
 1   name                  49437 non-null  object 
 2   homepage_url          45989 non-null  object 
 3   category_list         45477 non-null  object 
 4    market               45470 non-null  object 
 5    funding_total_usd    49438 non-null  object 
 6   status                48124 non-null  object 
 7   country_code          44165 non-null  object 
 8   state_code            30161 non-null  object 
 9   region                44165 non-null  object 
 10  city                  43322 non-null  object 
 11  funding_rounds        49438 non-null  float64
 12  founded_at            38554 non-null  object 
 13  founded_month         38482 non-null  object 
 14  founded_quarter       38482 non-null  object 
 15  founded_year       

> The columns capture various aspects such as company details, funding information, geographic location, and founding timeline.

>The dataset showcases a mix of data types, including object (16 columns) and float64 (23 columns). This variety allows for a comprehensive analysis of both categorical and numerical features.

>Important columns with a significant number of missing values include:   `permalink`, `name`, `homepage_url`, `category_list`, `market`, `status`, `country_code`, `state_code`, `region`, `city`, `founded_at`, `founded_month`, `founded_quarter`, and `founded_year`. Proper handling of missing data is necessary to ensure accurate analysis.

>Key numerical columns related to funding rounds and amounts include `funding_rounds`, `funding_total_usd`, and various funding round columns (`seed`, `venture`, `round_A`, etc.). These columns provide insights into the financial aspects of the startups.

>The dataset also includes temporal information such as `first_funding_at` and `last_funding_at`, which can be converted to datetime data type for time-based analysis.

>Overall, this dataset provides a rich foundation for advanced machine learning tasks, allowing for comprehensive exploration of startup characteristics and funding patterns.


## Check shape of dataframe

In [5]:
# Checking for the shape of the data.
desc.data_shape()

The DataFrame has:
	* 54294 rows
	* 39 columns 



## checking for the categorical columns

In [6]:
#checking for categorical columns.
categorical_columns = data.select_dtypes(include=['object', 'category'])
categorical_columns.columns

Index(['permalink', 'name', 'homepage_url', 'category_list', ' market ',
       ' funding_total_usd ', 'status', 'country_code', 'state_code', 'region',
       'city', 'founded_at', 'founded_month', 'founded_quarter',
       'first_funding_at', 'last_funding_at'],
      dtype='object')

These categorical columns provide important information about the company details, funding information, geographic location, and founding timeline of the startup companies.

Analyzing these columns can help gain insights into the industry focus, funding status, geographical distribution, and temporal aspects of the startup ecosystem represented in the dataset.

## checking for the ccontinuous  columns

In [7]:
#checking for numeric columns.
non_categorical_columns = data.select_dtypes(exclude=['object', 'category'])
non_categorical_columns.columns

Index(['funding_rounds', 'founded_year', 'seed', 'venture',
       'equity_crowdfunding', 'undisclosed', 'convertible_note',
       'debt_financing', 'angel', 'grant', 'private_equity', 'post_ipo_equity',
       'post_ipo_debt', 'secondary_market', 'product_crowdfunding', 'round_A',
       'round_B', 'round_C', 'round_D', 'round_E', 'round_F', 'round_G',
       'round_H'],
      dtype='object')

The non-categorical columns in the dataset consist of the following features:

1. `funding_rounds`: This column represents the number of funding rounds a startup has gone through. It provides insights into the funding activity of the companies.

2. `founded_year`: This column captures the year in which each startup was founded. It can be used to analyze trends and patterns related to the age of the companies.

3.`seed`, `venture`, `equity_crowdfunding`, `undisclosed`, `convertible_note`, `debt_financing`, `angel`, `grant`, `private_equity`, `post_ipo_equity`, `post_ipo_debt`, `secondary_market`, `product_crowdfunding`, `round_A`, `round_B`, `round_C`, `round_D`, `round_E`, `round_F`, `round_G`, `round_H`: These columns represent the amounts of funding raised in different funding rounds. They provide valuable information on the financial aspects of the startups.

## Concise statistical description of numeric features

In [8]:
# Obtaining statistical summary of the dataset
desc.data_describe()

funding_rounds  founded_year          seed       venture  \
count    49438.000000  38482.000000  4.943800e+04  4.943800e+04   
mean         1.696205   2007.359129  2.173215e+05  7.501051e+06   
std          1.294213      7.579203  1.056985e+06  2.847112e+07   
min          1.000000   1902.000000  0.000000e+00  0.000000e+00   
25%          1.000000   2006.000000  0.000000e+00  0.000000e+00   
50%          1.000000   2010.000000  0.000000e+00  0.000000e+00   
75%          2.000000   2012.000000  2.500000e+04  5.000000e+06   
max         18.000000   2014.000000  1.300000e+08  2.351000e+09   

       equity_crowdfunding   undisclosed  convertible_note  debt_financing  \
count         4.943800e+04  4.943800e+04      4.943800e+04    4.943800e+04   
mean          6.163322e+03  1.302213e+05      2.336410e+04    1.888157e+06   
std           1.999048e+05  2.981404e+06      1.432046e+06    1.382046e+08   
min           0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
25%           0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
50%           0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
75%           0.000000e+00  0.000000e+00      0.000000e+00    0.000000e+00   
max           2.500000e+07  2.924328e+08      3.000000e+08    3.007950e+10   

              angel         grant  ...  secondary_market  \
count  4.943800e+04  4.943800e+04  ...      4.943800e+04   
mean   6.541898e+04  1.628453e+05  ...      3.845592e+04   
std    6.582908e+05  5.612088e+06  ...      3.864461e+06   
min    0.000000e+00  0.000000e+00  ...      0.000000e+00   
25%    0.000000e+00  0.000000e+00  ...      0.000000e+00   
50%    0.000000e+00  0.000000e+00  ...      0.000000e+00   
75%    0.000000e+00  0.000000e+00  ...      0.000000e+00   
max    6.359026e+07  7.505000e+08  ...      6.806116e+08   

       product_crowdfunding       round_A       round_B       round_C  \
count          4.943800e+04  4.943800e+04  4.943800e+04  4.943800e+04   
mean           7.074227e+03  1.243955e+06  1.492891e+06  1.205356e+06   
std            4.282166e+05  5.531974e+06  7.472704e+06  7.993592e+06   
min            0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%            0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%            0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%            0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max            7.200000e+07  3.190000e+08  5.420000e+08  4.900000e+08   

            round_D       round_E       round_F       round_G       round_H  
count  4.943800e+04  4.943800e+04  4.943800e+04  4.943800e+04  4.943800e+04  
mean   7.375261e+05  3.424682e+05  1.697692e+05  5.767067e+04  1.423197e+04  
std    9.815218e+06  5.406915e+06  6.277905e+06  5.252312e+06  2.716865e+06  
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
max    1.200000e+09  4.000000e+08  1.060000e+09  1.000000e+09  6.000000e+08  

[8 rows x 23 columns]

Here are some important observations:

- The `funding_rounds` column has a mean value of approximately 1.70, indicating that, on average, startups have participated in around 1-2 funding rounds.

- The `founded_year` column has a mean value of approximately 2007, with a standard deviation of 7.58. This suggests that the startups in the dataset were founded between the early 1900s and 2014.

- The `seed`, `venture`, and other funding round columns (`round_A` to `round_H`) show varying mean values and standard deviations, indicating significant variations in funding amounts across different rounds.

- The `funding_total_usd` column is represented in scientific notation. The mean value is approximately 2.17e+05, with a standard deviation of 1.06e+06.

- All the numerical columns have a minimum value of 0, indicating that some startups may not have received any funding in certain rounds or types.

- The maximum values for the funding round columns vary significantly, indicating that some startups have secured substantial amounts of funding in specific rounds.

- It is important to note that the values in the `founded_year` column range from 1902 to 2014, with a large standard deviation. This suggests the presence of outliers or inconsistencies in the dataset. 

- the funding amounts across the different rounds vary widely, with some startups receiving zero funding in certain rounds.

 ## Checking for duplicated rows 

In [ ]:
# Checking for duplicates. 
def count_duplicates(data):
    duplicate_count = data.duplicated().sum()
    return duplicate_count
duplicate_count = count_duplicates(data)
print(duplicate_count)

Considering the high number of duplicate rows (4855), it is crucial to handle duplicates appropriately to avoid bias or inaccuracies in the analysis.

 ## Checking for missing values

In [ ]:
def plot_missing_values(data):
    missing_values = data.isnull().sum()
    percentage_missing = (missing_values / len(data)) * 100

    missing_data = pd.DataFrame({'Missing Values': missing_values, 'Percentage Missing': percentage_missing})
    print(missing_data)

    # Plot missing values
    ax = data.isna().sum().sort_values(ascending=False)[:15][::-1].plot(kind='barh', figsize=(9, 10))
    plt.title('Percentage of Missing Values Per Column', fontdict={'size': 15})
    for p in ax.patches:
        percentage = '{:,.0f}%'.format((p.get_width() / data.shape[0]) * 100)
        width, height = p.get_width(), p.get_height()
        x = p.get_x() + width + 0.02
        y = p.get_y() + height / 2
        ax.annotate(percentage, (x, y))
    
    plt.show()

# print the missing values:
plot_missing_values(data)

The dataset contains missing values in multiple columns. The highest percentage of missing values is observed in the `state_code` column (44.45%), followed by columns related to the startup's founding information. Other columns, including funding rounds and various attributes, also exhibit missing values ranging from 8.94% to 29.12%. 

Considering the varying percentages of missing values across different columns, it is crucial to carefully assess the impact of these missing values on the analysis and select appropriate strategies for handling them.

## Checking and Visualizing outliers 

In [ ]:
# Visualizing outliers. 

# Create box plots for numerical columns
plt.figure(figsize=(10, 8))
sns.boxplot(data=data)
plt.xticks(rotation=90)
plt.show()

Some data points fall significantly outside the normal range of values and can indicate potential anomalies or errors in the dataset.

Outliers may impact statistical analyses and machine learning models, leading to biased or inaccurate results.

# <span style="color: #008000;">DATA CLEANING</span>


To ensure the development of a robust and accurate model, several data cleaning techniques will be applied to the dataset. The following techniques will be employed:

1 `Completeness`: This technique involves addressing missing values within the dataset. Steps will be taken to identify and handle missing data appropriately, either through imputation or removal, to ensure that the dataset is complete.

2 `Consistency`: The consistency of the data will be examined to identify any discrepancies or irregularities. Inconsistencies in variables, such as conflicting formats or conflicting information within the dataset, will be addressed and resolved to maintain data integrity.

3 `Validity`: Validity refers to the accuracy and relevance of the data. Data validation techniques will be applied to verify that the values within each variable align with expected ranges or predefined criteria. Any invalid or erroneous data points will be rectified or removed from the dataset.

4 `Collinearity`: Collinearity examines the presence of highly correlated variables. Multicollinearity can affect the model's stability and interpretability. Techniques such as correlation analysis will be utilized to identify highly correlated variables. If necessary, appropriate measures, such as feature selection or dimensionality reduction, will be implemented to address collinearity issues.

5 `Outlier removal`: Outliers, which are extreme or unusual data points, can adversely impact the model's performance. Robust statistical techniques will be employed to detect and handle outliers effectively. Outliers may be removed or transformed to minimize their influence on the analysis and modeling process.

By applying these data cleaning techniques, the dataset will be refined and prepared, ensuring the reliability and accuracy of the data before proceeding with the subsequent stages of analysis and modeling.

## Renamimg column(s) to remove spaces.

By removing spaces in column names, the code improves the usability and consistency of the dataset. Column names without spaces are easier to work with and less prone to errors when referencing them in code or performing data manipulations.

In [ ]:
# Renamimg columns to remove spaces.
data.rename(columns={' funding_total_usd ':'funding_total_usd',' market ':'market'}, inplace=True)

## dropping duplicated rows.

This helps ensure data integrity and eliminates redundancy in the dataset. 

In [ ]:
#dropping duplicate rows.
data = data.drop_duplicates()
print(data.shape)

## Type Casting: Converting Funding Total to Float

In [ ]:
# Type casting
data['funding_total_usd'] = data['funding_total_usd'].apply(str).str.replace(',','')
data['funding_total_usd'] = data['funding_total_usd'].replace(' -   ',0).astype("float")


These type casting operations are useful for converting the 'funding_total_usd' column from a string representation to a numeric format, facilitating further analysis and computations involving funding amounts.

## dropping missing values.

In [ ]:
#dropping null values.
data = data.dropna(how="all")

In [ ]:
#Confirming for missing values.
missing_values = data.isnull().sum()
print(missing_values)

## Addressing Missing Values for Specific Columns
To handle missing values and further process specific columns, the following steps can be taken:

1. Convert the '**founded_year**' column to a float data type initially to handle NaN values.
2. After handling the NaN values, convert the '**founded_year**' column to an integer data type to ensure the values are represented correctly.
3. Cast the values in the '**funding_rounds**' column to the integer data type for consistency and numerical operations.
4. Convert the values in the '**first_funding_at**' and '**last_funding_at**' columns to datetime data type for better analysis and time-based operations.
5. Remove data entries with zero funds, as they may introduce biases or affect calculations. This step helps to ensure the accuracy of funding-related analysis.

In [ ]:
#Further dealing with missing values for specific columns.

def preprocess_data(data):
    # Convert founded_year column to float first to handle NaN values.
    data['founded_year'] = data['founded_year'].astype(float)

    # Casting the values in the 'funding_rounds' column to the integer data type.
    data['funding_rounds'] = data['funding_rounds'].astype(int)

    # Converting the values in the 'first_funding_at' and 'last_funding_at'
    # columns to datetime data type.
    data['first_funding_at'] = pd.to_datetime(data['first_funding_at'], errors='coerce')
    data['last_funding_at'] = pd.to_datetime(data['last_funding_at'], errors='coerce')

    # Remove NaN values
    data.dropna(inplace=True)

    # Convert founded_year column to integer after removing NaN values
    data['founded_year'] = data['founded_year'].astype(int)

    # Remove data with zero funds
    index_no_funding = data[data['funding_total_usd'] <= 0].index
    data.drop(index_no_funding, inplace=True)

    #Confirming for missing values.
    has_missing_values = data.isnull().any().any()
    

    return data

# print:
data = preprocess_data(data)

### Confirming there are no missing values

In [ ]:
# Confirming for missing values.
print(missing_values)

## Dealing with outliers

> Dealing with the extreme or unusual data points that can adversely impact the model's performance. 

Based on the empirical rule or the 68-95-99.7 rule. According to this rule, in a normal distribution, approximately 99.7% of the data falls within 3 standard deviations from the mean. By using a factor of 3, we are considering values that are relatively far from the mean as potential outliers.

In [ ]:
# Convert DataFrame to numeric type
data_numeric = data.apply(pd.to_numeric, errors='coerce')

# Calculate the upper and lower thresholds for outlier detection
upper_threshold = 3 * data_numeric.std() + data_numeric.mean()
lower_threshold = data_numeric.mean() - 3 * data_numeric.std()


# Replace the outliers with the column's median
data_numeric = data_numeric.apply(lambda x: x.where(~((x > upper_threshold[x.name]) | (x < lower_threshold[x.name]))), axis=0)

# Check if any outliers are still present
outliers_present = (data_numeric > upper_threshold) | (data_numeric < lower_threshold)

# Return True if any outliers are found, False otherwise
data_contains_outliers = outliers_present.any().any()
print(data_contains_outliers)


# <span style="color: #008000;">EXPLORATORY DATA ANALYSIS(EDA)</span>

## <span style ='color:cyan'> Univariate Analysis.  </span>

### Analysis on `status` Feature
`status` will be used as the dependent variable in this analysis.

Understanding the distribution of values in the `status` column is essential for analyzing the state of the startups in the dataset. It provides insights into the current status of the startups, such as whether they are actively operating, have been acquired, or have closed down.

In [ ]:
#Importing libraries for visualization.
import plotly.express as px

# Count the occurrences of each status
status_counts = data['status'].value_counts()

# Create a bar plot using Plotly with different colors
fig = px.bar(x=status_counts.index, y=status_counts.values, color=status_counts.index)

# Set the plot title and axis labels
fig.update_layout(title="Status Counts", xaxis_title="Status", yaxis_title="Count")

# Display the plot
fig.show()

The countplot suggests that the majority of the startups in the dataset are classified as 'operating', followed by a smaller number of 'acquired' startups and a relatively smaller number of 'closed' startups.

### top 10% `total_funding` startups.¶

In [ ]:
top_fund=(data.funding_total_usd.sort_values(ascending=False)[1:5000].sum())/(data.funding_total_usd.sort_values(ascending=False).sum())

plt.figure()
cmap = plt.get_cmap("tab20c")
outer_colors = cmap(np.arange(2)*13)
plt.pie([top_fund,1-top_fund], radius=2, colors=outer_colors,labels=['Top 10%','Others'],autopct='%1.1f%%',
       wedgeprops=dict(width=0.5, edgecolor='w'))
plt.title('Top 10% Companies\' total funding',pad=100)
plt.show()

We got over 50k companies here, top 10% gets most of the funding.
It's a winner takes all game

### What is the distribution of founded years?


In [ ]:
# Create a figure with specified size
plt.figure(figsize=(10, 6))

# Create a histogram plot using seaborn
sns.histplot(data=data, x='founded_year', color='skyblue')

# Set x-axis label
plt.xlabel('Founded Year')

# Set y-axis label
plt.ylabel('Count')

# Set plot title
plt.title('Distribution of Founded Year')

# Display the plot
plt.show()

>The distribution of founded years exhibits a concentration of startups from the 2000s to the present period, as indicated by the prominent peaks or bars in the histogram. This suggests a significant surge in entrepreneurial activity and startup formations during this timeframe.

>In contrast, there is a relatively lower number of startups founded between the 1980s and 2000. While the frequency of startups increases during this period, it is not as pronounced as the recent years.

>Furthermore, the histogram shows a notable decrease in the number of startups founded before the 1980s. Very few startups were established during this early period, indicating a less developed or nascent startup ecosystem.

### What are the Top 10 Country Code Counts?

In [ ]:
# Get the value counts for the "country_code" column
country_counts = data["country_code"].value_counts()[:10]

# Create a bar plot using Plotly with different colors
fig = px.bar(x=country_counts.index, y=country_counts.values, color=country_counts.index)

# Set the plot title and axis labels
fig.update_layout(title="Top 10 Country Code Counts", xaxis_title="Country Code", yaxis_title="Count")

# Rotate x-axis labels for better readability
fig.update_xaxes(tickangle=50)

# Display the plot
fig.show()

USA has the most number of startups

### What are the Top 10 City Code Counts?

In [ ]:
# Get the value counts for the "city" column
city_counts = data["city"].value_counts()[:10]

# Create a bar plot using Plotly with different colors
fig = px.bar(x=city_counts.index, y=city_counts.values, color=city_counts.index)

# Set the plot title and axis labels
fig.update_layout(title="Top 10 City Counts", xaxis_title="City", yaxis_title="Count")

# Rotate x-axis labels for better readability
fig.update_xaxes(tickangle=50)

# Increase the figure size
fig.update_layout(width=800, height=600)

# Display the plot
fig.show()

San Francisco have the highest no. of startup companies than other cities.

### What is the distribution of funding rounds?


In [ ]:
# Create a figure with specified size
plt.figure(figsize=(10, 6))

# Create a countplot using seaborn
sns.countplot(data=data, x='funding_rounds')

# Set x-axis label
plt.xlabel('Funding Rounds')

# Set y-axis label
plt.ylabel('Count')

# Set plot title
plt.title('Distribution of Funding Rounds')

# Display the plot
plt.show()

Most company only get one round of funding.


### What are the top 10 markets based on count?


In [ ]:
# Get the top 10 markets by count
top_markets = data['market'].value_counts().nlargest(10)

# Create a figure with specified size
plt.figure(figsize=(10, 6))

# Create a pie chart
plt.pie(top_markets, labels=top_markets.index, autopct='%1.1f%%')

# Set plot title
plt.title('Top 10 Markets')

# Display the plot
plt.show()

### 'Markets that most start-ups closed

In [ ]:
top_markets=(data.market.value_counts()[data.market.value_counts()>24]).index
# We chose top markets to analyze,which are greater than 24 startups in that market.
sub_data=data[data.market.isin(top_markets)]

status_market_1=sub_data.groupby('market')['status'].value_counts(normalize=True).to_frame()
status_market_1.columns=['prop']
status_market_2=sub_data.groupby('market')['status'].value_counts().to_frame()
status_market_2.columns=['quantity']
status_market=pd.concat([status_market_1,status_market_2],axis=1)
status_market.reset_index()
status_market_closed=status_market.query('status=="closed"').reset_index(level=1)
status_market_closed.sort_values(by='prop',ascending=False)[0:10].style.background_gradient(cmap='Reds',subset=['prop']).set_caption('Markets that most start-ups closed')

### Markets that least start-ups closed

In [ ]:
status_market_closed.sort_values(by='prop',ascending=True)[0:10].style.background_gradient(cmap='Greens_r',subset=['prop']).set_caption('Markets that least start-ups closed')

## <span style ='color:cyan'> Multivariate Analysis.  </span>

### What are the funding totals by country?


In [ ]:
# Calculate funding total by country
funding_by_country = data.groupby('country_code')['funding_total_usd'].sum().sort_values(ascending=False)

# Create a bar plot using Plotly
fig = px.bar(x=funding_by_country.index, y=funding_by_country.values)

# Set the plot title and axis labels
fig.update_layout(title='Funding Total by Country', xaxis_title='Country', yaxis_title='Funding Total (USD)')

# Rotate x-axis labels for better readability
fig.update_xaxes(tickangle=90)

# Adjust the width of the figure
fig.update_layout(width=700)

# Display the plot
fig.show()

### What are the top 10 funding totals by founded year? 

In [ ]:
# Calculate the top 10 funding totals by founded year
top_10_funding = data.groupby('founded_year')['funding_total_usd'].sum().sort_values(ascending=False)[:10]

# Define a color palette
color_palette = ['red', 'orange', 'yellow', 'green', 'blue', 'purple', 'pink', 'brown', 'gray', 'cyan']

# Create the bar plot with the specified color palette
plt.figure(figsize=(10, 6))
top_10_funding.plot(kind='bar', color=color_palette)

# Set y-axis label
plt.ylabel('Total Funding in USD')

# Set plot title
plt.title('Top 10 Funding over the Years')

# Display the plot
plt.show()

### What is the average funding rounds by status?


In [ ]:
# Calculate the average funding rounds by status
funding_rounds_by_status = data.groupby('status')['funding_rounds'].mean().sort_values(ascending=False)

# Create a bar plot using seaborn
plt.figure(figsize=(8, 6))
sns.barplot(x=funding_rounds_by_status.index, y=funding_rounds_by_status.values)

# Set x-axis label
plt.xlabel('Status')

# Set y-axis label
plt.ylabel('Average Funding Rounds')

# Set plot title
plt.title('Funding Rounds by Status')

# Display the plot
plt.show()

### What is the relationship between funding total and funding rounds by market?


In [ ]:
# Calculate the mean funding rounds by market
market_funding_rounds = data.groupby('market')['funding_rounds'].mean()

# Calculate the total funding by market
market_funding_total = data.groupby('market')['funding_total_usd'].sum()

# Create a bubble plot using Plotly Express
fig = px.scatter(x=market_funding_rounds, y=market_funding_total, color=market_funding_total)

# Set the plot title and axis labels
fig.update_layout(
    title="Bubble Plot of Funding Total vs. Funding Rounds by Market",
    xaxis_title="Funding Rounds (mean)",
    yaxis_title="Funding Total (sum)"
)

# Adjust the width of the figure
fig.update_layout(width=900)

# Display the plot
fig.show()

### What is the total funding received by companies?

In [ ]:
# Calculate the total funding received by companies
total_funding = data.groupby('name')['funding_total_usd'].sum().sort_values(ascending=False)

# Create a figure with a specified size
plt.figure(figsize=(12, 6))

# Create a bar plot using seaborn
sns.barplot(x=total_funding.index[:10], y=total_funding.values[:10])

# Set x-axis label
plt.xlabel('Company')

# Set y-axis label
plt.ylabel('Total Funding (USD)')

# Set plot title
plt.title('Total Funding Received by Companies - Top 10')

# Rotate x-axis labels for better readability
plt.xticks(rotation=90)

# Display the plot
plt.show()

# <span style="color: #008000;">FEATURE ENGINEERING/SELECTION</span>

# <span style="color: #008000;">MODELLING</span>

- transformation 
- imbalance
- correlation 
- different handles missing values

# <span style="color: #008000;">EVALUATION</span>

# <span style="color: #008000;">RECOMMENDATIONS </span>

# <span style="color: #008000;">CONCLUSION</span>

### The Top Countries in Terms of Startup Number

In [ ]:
#There are some startups their countr info is unknown i made them 'others'
def make_other(value):
    if str(value) == str(np.nan):
        return 'others'
    else:
        return str(value)
    
df['country_code'] = df['country_code'].apply(make_other)
df['country_code'].value_counts(dropna=False)    

In [ ]:
fig = px.pie(df,names=df.country_code,
             title = "The Top Countries in Terms of Startup Number",
             color_discrete_sequence=px.colors.sequential.RdBu,
             hole = 0.15,
             width =800,
             height =600)

fig.update_layout(uniformtext_minsize=10, uniformtext_mode='hide', title_x=0.1)
fig.update_traces(textposition='inside', textinfo='percent+label')

fig.show()

### most funded categoriess

In [ ]:
import wordcloud

%%time
category_list=data.category_list.str.split('|').to_list()
categ=[]

for i in category_list:
    if isinstance(i,list):
        categ=categ+i
categ=list(filter(None,categ))
categ=' '.join(categ)

w=wordcloud.WordCloud(background_color='White',scale=32)

fig = plt.figure(figsize=(12, 8))
w.generate(categ)
plt.axis('off')
plt.imshow(w)

### Top 10 Market of Start-ups

In [ ]:
market_con=data.groupby('country_code')['market'].value_counts().to_frame()

In [ ]:
fig,axes=plt.subplots(2,2,figsize=(20,10))
cont=['USA','GBR','CAN','CHN']
for con,ax in zip(cont,range(1,5)):
    ax=plt.subplot(2,2,ax)
    ax=sns.barplot(y=market_con.loc[con].index[0:10],x=market_con.loc[con].market[0:10],orient='h',palette='Blues_r')
    ax.set_xlabel('Market Counts In {}'.format(con))
plt.suptitle('Top 10 Market of Start-ups',size=40)
plt.show()

### What is the relationship between funding and founded year? (With points colored by status) 


In [ ]:
# Create a figure with specified size
plt.figure(figsize=(10, 6))

# Create a scatter plot using seaborn
sns.scatterplot(data=data, x='founded_year', y='funding_total_usd', hue='status', palette='Set1')

# Set x-axis label
plt.xlabel('Founded Year')

# Set y-axis label
plt.ylabel('Funding in USD')

# Set plot title
plt.title('Funding vs Founded Year (by Status)')

# Add legend
plt.legend()

# Display the plot
plt.show()

### The Companies get more than 10 Million Dollars by angel investors

In [ ]:
angels = df[df['angel']>=10000000]

fig = px.bar(angels, 
             x=angels['name'], 
             y=angels['angel'],
             height=800,
             color=angels['angel'],
             title = 'The Companies get more than 10M($) by angel',
             text=angels['name'],
             color_continuous_scale='greens',
             )
fig.update_traces(dict(marker_line_width=0))
fig.update_layout(title_x = 0.5)
fig.update_xaxes(tickangle=90, title='Companies')
fig.update_yaxes(title='invest amount by angel($)')
fig.update_layout(barmode='stack', xaxis={'categoryorder':'total descending'})
fig.show()

### Number of Startups Country wise on Map

In [ ]:
countries = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/2014_world_gdp_with_codes.csv')
countries = countries.drop('GDP (BILLIONS)', 1)
countries = countries.rename(columns={'CODE': 'country_code'})

In [ ]:
countries

In [ ]:
fig_country_map = go.Figure()
fig_country_map.add_trace(go.Choropleth(locations=df['country_code'].value_counts().index,
                                       z=df['country_code'].value_counts(),
                                       colorscale='Peach', 
                                       colorbar_title="Nos. of Startups founded"))
fig_country_map.update_layout(title_text="Number of Startups Country wise.", title_x=0.5, title_font_size=20)

fig_country_map.show()